pkg: new package for handling byteorder

aalemayhu opened the pull request:
Closes: #285 (Micro: Fix `common.Swab16` with proper `encoding/binary` usage)

tgraf commented on Makefile.defs:
```diff
@@ -5,9 +5,10 @@ CONFDIR?=/etc
 
 INSTALL = install
 
+NATIVE_BYTE_ORDER=$(shell lscpu | grep Endian |awk '{ print $$3"_"$$4 }')
```
`lscpu` may not be available everywhere, what about we use something like this instead?
```
const INT_SIZE int = int(unsafe.Sizeof(0))
var i int = 0x1
bs := (*[INT_SIZE]byte)(unsafe.Pointer(&i))
if bs[0] == 0 {
    // big endian
} else {
    // little endian
}
```

tgraf commented on pkg/byteorder/byteorder.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+// Swab16 swaps the endianness of n.
+func Swab16(n uint16) uint16 {
```
Instead of this, it would be nice to have `ntohs` and `htons` equivalents instead to document where we are converting from and to network byte order.

aalemayhu replied:
Where do we need to use `ntohs `, looks like there is only `htons` currently.
What am I missing?

tgraf replied:
After reading from maps and all occurrences in the monitor where we print values which are reported by the monitor in network byte order.

tgraf replied:
There is currently a `Convert()` function for `bpf.Map` which must be split in two as `Convert()` is used for both, from and to network.

aalemayhu replied:
Hmm, think I might have missed this Convert part. Will go through and see.

tgraf replied:
`Convert()` should be split into `ToHost()` after reading from the kernel and `ToNetwork()` before writing to the kernel.

tgraf commented on cilium/cmd/monitor.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/cilium/cilium/pkg/bpf"
 	"github.com/cilium/cilium/pkg/bpfdebug"
 
+	"github.com/cilium/cilium/pkg/byteorder"
```
This should be moved to the cilium imports block

tgraf replied:
The entire imports block should look like this:
```
import (
 	"bytes"
 	"encoding/binary"
 	"fmt"
 	"os"
 	"os/signal"
 	"runtime"
  
  	"github.com/cilium/cilium/pkg/bpf"
  	"github.com/cilium/cilium/pkg/bpfdebug"
  	"github.com/cilium/cilium/pkg/byteorder"

 	"github.com/spf13/cobra"
  	"golang.org/x/sys/unix"
  )

tgraf replied:
This still needs to be addressed.

tgraf commented on daemon/daemon.go:
```diff
@@ -405,19 +405,19 @@ func (d *Daemon) init() error {
 
 	ipv4GW := d.conf.NodeAddress.IPv4Address
-	fmt.Fprintf(fw, "#define IPV4_GATEWAY %#x\n", binary.LittleEndian.Uint32(ipv4GW))
+	fmt.Fprintf(fw, "#define IPV4_GATEWAY %#x\n", byteorder.NetworkToHostU32(ipv4GW))
```
All of these seem reversed. We want to write the `IPV4_GATEWAY` et al in network byte order.

borkmann replied:
I think above conversion is not correct. BPF code treats this as network byte order, so this would break things.

Take `fmt.Fprintf(fw, "#define LXC_IPV4 %#x\n", binary.BigEndian.Uint32(e.IPv4))`, for example. That is stored in host order, hence `bpf_htonl()` is needed in BPF code.

I think this `binary.BigEndian`/`binary.LittleEndian` interface is highly confusing.

tgraf replied:
Agreed that the golang API is very confusing. We had several bugs because of this. @scanf is  now trying to improve this. ntohs/htons is much easier to understand.

aalemayhu replied:
Should this one line be fixed separately first, if it hasn't already in master(need to check)?

Currently in the process of slicing the patch up into a series where more tests are added.
Will respin with feedback addressed later today.

tgraf commented on daemon/monitor.go:
```diff
@@ -21,19 +21,20 @@ package main
 	"time"
 
+	"bytes"
+	"encoding/binary"
```
We want to keep native golang imports and cilium import blocks separated.

aalemayhu committed:
byteorder: add documentation file for byteorder package

this should show a synopsis on godoc[0].

[0]: https://godoc.org/github.com/cilium/cilium

Closes: #285 (Micro: Fix common.Swab16 with proper encoding/binary usage)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -161,7 +161,7 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	fw.WriteString(common.FmtDefineAddress("LXC_IP", e.IPv6))
 	if e.IPv4 != nil {
-		fmt.Fprintf(fw, "#define LXC_IPV4 %#x\n", binary.BigEndian.Uint32(e.IPv4))
+		fmt.Fprintf(fw, "#define LXC_IPV4 %#x\n", byteorder.HostToNetworkSliceU32(e.IPv4))
```
This would be correct in general but the BPF program now does, htonl since #747. The BPF side would need to change at the same time.

jrfastab replied:
Well even before that PR it was doing  mix of htonl and not so it needs to be made consistent one way or the other. I don't mind if the BPF sidie does the conversion or the writeHeaderFile as long as it is done once.

aalemayhu replied:
@tgraf If I make that change then the connectivity tests start failing, not sure why but looking into it 
https://github.com/cilium/cilium/blob/master/tests/01-ct.sh#L99-L102

tgraf replied:
Did you rebase on top of #747?

aalemayhu replied:
Rebased on master which includes 697b6d18ce17 (bpf: fix drop due to wrong l3/l4 csum in ipv4_redirect_to_host_port, 2017-05-17)

tgraf replied:
The BPF program in master expects `LXC_IPV4` to be in host byte order at the moment. Let's keep it as-is for this PR and we can optimize this later. The test is failing because you would also have to change the bpf side first.

tgraf commented on pkg/bpfdebug/debug.go:
```diff
@@ -131,14 +129,14 @@ func ctInfo(arg1 uint32, arg2 uint32) string {
 func verdictInfo(arg uint32) string {
-	proxyPort := common.Swab16(uint16(arg >> 16))
-	revnat := common.Swab16(uint16(arg & 0xFFFF))
+	proxyPort := byteorder.HostToNetworkU16(uint16(arg >> 16))
```
These all need to be `NetworkToHost`

aalemayhu commented on pkg/maps/ctmap/ipv4.go:
```diff
@@ -99,15 +107,22 @@ func (k *CtKey4Global) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
+	return &n
+}
+
 func (k *CtKey4Global) String() string {
```
mixed up names, will fix.

tgraf commented on cilium/cmd/monitor.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/cilium/cilium/pkg/bpf"
 	"github.com/cilium/cilium/pkg/bpfdebug"
 	"github.com/cilium/cilium/pkg/byteorder"
+
```
Can you squash this changes into the commits where the import order was wrong?

tgraf commented on daemon/daemon.go:
```diff
@@ -413,19 +413,19 @@ func (d *Daemon) init() error {
 
 	ipv4GW := d.conf.NodeAddress.IPv4Address
-	fmt.Fprintf(fw, "#define IPV4_GATEWAY %#x\n", binary.LittleEndian.Uint32(ipv4GW))
+	fmt.Fprintf(fw, "#define IPV4_GATEWAY %#x\n", byteorder.NetworkToHostSliceU32(ipv4GW))
```
I'm getting confused how all of this can work. IPV4_GATEWAY must be written in network byte order.

aalemayhu committed:
tests: check printed container ids

Related-to: #746 (container id in BPF debug message is wrong)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
bpf: don't call bpf_htonl for LXC_IPV4

It's should already be in network byte order from Go side.

Related-to: #285 (Micro: Fix `common.Swab16` with proper `encoding/binary` usage)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aanm commented on cilium/cmd/monitor.go:
```diff
@@ -110,7 +111,7 @@ func match(messageType int, src uint16, dst uint32) bool {
 	dn := bpfdebug.DropNotify{}
 
-	if err := binary.Read(bytes.NewReader(data), binary.LittleEndian, &dn); err != nil {
+	if err := binary.Read(bytes.NewReader(data), byteorder.Native, &dn); err != nil {
```
`Native`? In what sense?

aalemayhu replied:
In the host endianness.

aanm commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -48,9 +48,8 @@ type CtKey interface {
-	// if necessary.
-	Convert() CtKey
+	// ToNetwork converts fields to network byte order.
+	ToNetwork() CtKey
```
Don't we need to convert it back to host by order? (For example, reading from maps and printing in go)

aalemayhu replied:
Yes, this was lost during rebasing. Will re add them.

aalemayhu committed:
pkg: introduce new package for handling byteorder

netlink is used for the byte order detection, which is already a
dependency available in the vendor directory. The package provides the
native byte order in the variable Native and wraps all byte order
specific reads and write call sites.

Related-to: #285 (Micro: Fix `common.Swab16` with proper `encoding/binary` usage)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

borkmann commented on pkg/byteorder/byteorder.go:
```diff
@@ -0,0 +1,122 @@
+	case uint16:
+		return Native.Uint16(nl.Uint16Attr(n.(uint16)))
+	case uint32:
+		return Native.Uint32(nl.Uint32Attr(n.(uint32)))
```
Why this detour and not using nl.SwapX() API (which is admittedly a very confusing API name in the sense that it only swaps when in little-endian)?

aalemayhu replied:
Not sure. Will fix and push out later with some other queued things.
Just need to wait for the local tests first. Thanks.
